File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-2.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights or not heights[0]:
            return []

        rows, cols = len(heights), len(heights[0])
        pacific = set()
        atlantic = set()

        def dfs(r, c, visit):
            if (r, c) in visit:
                return
            visit.add((r, c))

            directions = [(1,0), (-1,0), (0,1), (0,-1)]
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if (
                    0 <= nr < rows and
                    0 <= nc < cols and
                    heights[nr][nc] >= heights[r][c]
                ):
                    dfs(nr, nc, visit)

        # Pacific: top row and left column
        for c in range(cols):
            dfs(0, c, pacific)
            dfs(rows - 1, c, atlantic)

        for r in range(rows):
            dfs(r, 0, pacific)
            dfs(r, cols - 1, atlantic)

        res = []
        for r in range(rows):
            for c in range(cols):
                if (r, c) in pacific and (r, c) in atlantic:
                    res.append([r, c])

        return res

File: Data Structures & Algorithms/valid-sudoku/submission-2.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<int> rows(9);
        vector<int> cols(9);
        vector<int> blocks(9);
        int ROWS = board.size();
        int COLS = board[0].size();
        for (int r=0; r<ROWS; r++) {
            for (int c=0; c<COLS; c++) {
                if (board[r][c] == '.') continue;
                int n = board[r][c] - '0';
                int b = (r/3)*3 + (c/3);
                if (rows[r] & (1<<n) || 
                    cols[c] & (1<<n) || 
                    blocks[b] & (1<<n)) return false;
                rows[r] |= (1<<n);
                cols[c] |= (1<<n);
                blocks[b] |= (1<<n);
            }
        }
        return true;
    }
};
